store/store.go

```diff
@@ -1191,6 +1213,116 @@ func (s *Store) Find(ctx context.Context, search *Search, user *auth.UserState)
 		return nil, ErrBadQuery
 	}
 
+	q := url.Values{}
+	q.Set("fields", strings.Join(s.searchFields, ","))
+	q.Set("architecture", s.architecture)
+
+	if search.Private {
+		q.Set("private", "true")
+	}
+
+	if search.Prefix {
+		q.Set("name", searchTerm)
+	} else {
+		if search.CommonID != "" {
+			q.Set("common-id", search.CommonID)
+		}
+		if searchTerm != "" {
+			q.Set("q", searchTerm)
+		}
+	}
+
+	// section is "category" in search v2
+	if search.Section != "" {
+		q.Set("category", search.Section)
+	}
+
+	// with search v2 all risks are searched by default (same as scope=wide
+	// with v1) so we need to restrict channel if scope is not passed.
+	if search.Scope == "" {
+		q.Set("channel", "stable")
+	} else if search.Scope != "wide" {
+		return nil, ErrInvalidScope
+	}
+
+	if release.OnClassic {
+		q.Set("confinement", "strict,classic")
+	} else {
+		q.Set("confinement", "strict")
+	}
+
+	u := s.endpointURL(searchEndpPathV2, q)
+	reqOptions := &requestOptions{
+		Method:   "GET",
+		URL:      u,
+		Accept:   jsonContentType,
+		APILevel: apiV2Endps,
+	}
+
+	var searchData searchV2Results
+	doRequest := func() (*http.Response, error) {
+		return s.doRequest(ctx, s.client, reqOptions, user)
+	}
+	readResponse := func(resp *http.Response) error {
+		ok := (resp.StatusCode == 200 || resp.StatusCode == 201)
+		// always decode on success; decode failures only if body is not empty
+		if !ok && resp.ContentLength == 0 {
+			return nil
+		}
+		ct := resp.Header.Get("Content-Type")
+		if ct != jsonContentType {
+			return nil
```

Comment:
this doesn't look right, if the answer is ok we should error if the body is not JSON. Also maybe we can accomplish this with:

```
resp, err := s.retryRequestDecodeJSON(ctx, reqOptions, user, &searchData, &searchData)
```

?   this passes searchData both as succes and failure destinations

Reply:
You're right, corrected the ok case.

Regarding the suggestion, I tried that and the problem is that this affects existing find tests that respond with error codes and plain text (due to attempted json decoding retryRequestDecodeJSON). Fixing this by handling content type (per XXX comment I made above in decodeJSONBody) breaks many other tests (unrelated to find) as they often send JSON without setting content-type; it's not clear whether these are test issues, but probably changing this in such a central place right now would be too big and risky change (therefore I opted for the specialized retry here).

Reply:
ok,  I see, I understand, we do want to target this to 2.44.  But please let's have a TODO about fixing this (hopefully in 2.45 or 2.46), because there's already enough messiness in store.go.

Reply:
Added a TODO.